src/s3/async_put_object.rs

```diff
@@ -0,0 +1,101 @@
+use aws_sdk_s3::error::PutObjectError;
+use aws_sdk_s3::model::ObjectCannedAcl;
+use aws_sdk_s3::output::PutObjectOutput;
+use aws_sdk_s3::SdkError;
+use futures::io::{Error, ErrorKind};
+use futures::task::{Context, Poll};
+use futures::{AsyncWrite, Future};
+use std::pin::Pin;
+
+use crate::s3::Client;
+
+enum PutObjectState<'a> {
+    Writing,
+    Closing(Pin<Box<dyn Future<Output = Result<PutObjectOutput, SdkError<PutObjectError>>> + 'a>>),
+    Closed,
+}
+
+pub struct AsyncPutObject<'a> {
+    client: &'a Client,
+    key: String,
+    bucket: String,
+    buf: Vec<u8>,
+    state: PutObjectState<'a>,
+}
+
+impl<'a> AsyncPutObject<'a> {
+    pub fn new(client: &'a Client, key: &str, bucket: &str) -> Self {
+        AsyncPutObject {
+            client,
+            key: key.into(),
+            bucket: bucket.into(),
+            buf: vec![],
+            state: PutObjectState::Writing,
+        }
+    }
+}
+
+impl<'a> AsyncWrite for AsyncPutObject<'a> {
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, Error>> {
+        match self.state {
+            PutObjectState::Writing => {
+                self.buf.extend(buf);
+                Poll::Ready(Ok(buf.len()))
+            }
+            _ => Poll::Ready(Err(Error::new(ErrorKind::Other, "bad poll_write"))),
+        }
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Result<(), Error>> {
+        match self.state {
+            PutObjectState::Writing => Poll::Ready(Ok(())),
+            _ => Poll::Ready(Err(Error::new(ErrorKind::Other, "bad poll_flush"))),
+        }
+    }
+
+    fn poll_close<'b>(
+        mut self: Pin<&'b mut AsyncPutObject<'a>>,
+        cx: &'b mut Context<'_>,
+    ) -> Poll<Result<(), Error>> {
+        match self.state {
+            PutObjectState::Writing => {
+                // Create the put_object future and transition to
+                // the Closing state.
+                let fut = self
+                    .client
+                    .put_object()
+                    .bucket(&self.bucket)
+                    .key(&self.key)
+                    .body(self.buf.clone().into())
```

Comment:
To avoid copying the bytes here, you could use [`std::mem::take`](https://doc.rust-lang.org/stable/std/mem/fn.take.html) to steal the buffer from `self.buf` (leaving an empty one in its place).